src/Workspaces/Core/MSBuild/MSBuild/ProjectFile/CommandLineArgumentReader.cs

```diff
@@ -0,0 +1,265 @@
+
+        protected CommandLineArgumentReader(MSB.Execution.ProjectInstance project)
+        {
+            Project = project;
+            _builder = ImmutableArray.CreateBuilder<string>();
+        }
+
+        protected abstract void ReadCore();
+
+        protected void Add(string name)
+        {
+            _builder.Add($"/{name}");
+        }
+
+        protected void Add(string name, string value)
+        {
+            _builder.Add($"/{name}:{value}");
+        }
+
+        protected void Add(string name, int value)
+        {
+            _builder.Add($"/{name}:{value}");
+        }
+
+        protected void AddIfNotNullOrWhiteSpace(string name, string value)
+        {
+            if (!string.IsNullOrWhiteSpace(value))
+            {
+                _builder.Add($"/{name}:{value}");
+            }
+        }
+
+        protected void AddIfTrue(string name, bool condition)
+        {
+            if (condition)
+            {
+                _builder.Add($"/{name}");
+            }
+        }
+
+        protected void AddIfTrue(string name, string value, bool condition)
+        {
+            if (condition)
+            {
+                _builder.Add($"/{name}:{value}");
+            }
+        }
+
+        protected void AddIfFalse(string name, bool condition)
+        {
+            if (!condition)
+            {
+                _builder.Add($"/{name}");
+            }
+        }
+
+        protected void AddQuoted(string name, string value)
+        {
+            _builder.Add($"/{name}:\"{value}\"");
+        }
+
+        protected void AddWithPlusOrMinus(string name, bool condition)
+        {
+            if (condition)
+            {
+                Add($"/{name}+");
+            }
+            else
+            {
+                Add($"/{name}-");
+            }
+        }
+
+        protected string GetDocumentFilePath(MSB.Framework.ITaskItem documentItem)
+        {
+            return GetAbsolutePath(documentItem.ItemSpec);
+        }
+
+        protected string GetAbsolutePath(string path)
+        {
+            var directoryPath = PathUtilities.GetDirectoryName(Project.FullPath);
+            return Path.GetFullPath(FileUtilities.ResolveRelativePath(path, directoryPath) ?? path);
+        }
+
+        protected void ReadAdditionalFiles()
+        {
+            var additionalFiles = Project.GetAdditionalFiles();
+            if (additionalFiles != null)
+            {
+                foreach (var additionalFile in additionalFiles)
+                {
+                    AddQuoted("additionalfile", GetDocumentFilePath(additionalFile));
+                }
+            }
+        }
+
+        protected void ReadAnalyzers()
+        {
+            var analyzers = Project.GetAnalyzers();
+            if (analyzers != null)
+            {
+                foreach (var analyzer in analyzers)
+                {
+                    Add("analyzer", GetDocumentFilePath(analyzer));
```

Comment:
FWIW, there were two because I was matching the style in the MSBuildTask. Now that I'm unifying and quoting only when the value is null, empty, or contains whitespace, this will be fixed.